xapian-core/diversify/diversify.cc

```diff
@@ -0,0 +1,195 @@
+{
+    LOGCALL(API, vector<Point>, "Diversify::compute_diff_dmset", dmset);
+    vector<Point> diff_dmset;
+    for (auto point : points) {
+	bool found_point = false;
+	Xapian::docid point_docid = point.get_document().get_docid();
+	for (auto doc : dmset) {
+	    if (point_docid == doc.get_document().get_docid()) {
+		found_point = true;
+		break;
+	    }
+	}
+
+	if (!found_point) {
+	    diff_dmset.push_back(point);
+	}
+    }
+
+    return diff_dmset;
+}
+
+double
+Diversify::evaluate_dmset(const vector<Point> &dmset)
+{
+    LOGCALL(API, double, "Diversify::evaluate_dmset", dmset);
+    double score_1 = 0, score_2 = 0;
+
+    for (auto doc : dmset)
+	score_1 += weights[doc.get_document().get_docid()];
+
+    vector<Point> diff_dmset = compute_diff_dmset(dmset);
+
+    for (auto point : diff_dmset) {
+	double min_dist = numeric_limits<double>::max();
+	unsigned int pos = 1;
+	for (auto doc : dmset) {
+	    auto key = make_pair(point.get_document().get_docid(),
+				 doc.get_document().get_docid());
+	    double sim = pairwise_sim[key];
+	    double weight = 2 * b * sigma_sqr * (1 / log(1 + pos)) * (1 - sim);
+	    min_dist = min(min_dist, weight);
+	    ++pos;
+	}
+	score_2 += min_dist;
+    }
+
+    return -lambda * score_1 + (1 - lambda) * score_2;
+}
+
+DocumentSet
+Diversify::get_dmset(const MSet &mset)
+{
+    LOGCALL(API, MSet, "Diversify::get_dmset", mset);
+    // Return original mset if no need to diversify
+    if (k == 0) {
+	DocumentSet dmset;
+	for (MSetIterator it = mset.begin(); it != mset.end(); ++it)
+	    dmset.add_document(it.get_document());
+	return dmset;
+    }
+
+    if (k > mset.size())
+	k = mset.size();
+
+    initialise_points(mset);
+    compute_similarities();
+    vector<Point> main_dmset(points.begin(), points.begin() + k);
+    vector<Point> curr_dmset = main_dmset;
+
+    while (true) {
+	bool found_better_dmset = false;
+	for (unsigned int i = 0; i < main_dmset.size(); ++i) {
+	    auto curr_doc = main_dmset[i];
+	    unsigned int best_score = evaluate_dmset(curr_dmset);
+
+	    vector<Point> diff_dmset = compute_diff_dmset(curr_dmset);
+
+	    bool found_better_doc = false;
+	    for (unsigned int j = 0; j < diff_dmset.size(); ++j) {
+		vector<Point> temp_dmset = curr_dmset;
+		temp_dmset[i] = diff_dmset[j];
+		unsigned int score = evaluate_dmset(temp_dmset);
+		if (score < best_score) {
+		    curr_doc = temp_dmset[i];
+		    best_score = score;
+		    found_better_doc = true;
+		}
+	    }
+	    if (found_better_doc) {
+		curr_dmset[i] = curr_doc;
+		found_better_dmset = true;
+	    }
+	}
+
+	// Terminate algorithm when there's no change in current
+	// document matchset
+	if (!found_better_dmset)
+	    break;
+
+	main_dmset = curr_dmset;
+    }
+
+    // Merge main_dmset and diff_dmset into final dmset
+    DocumentSet dmset;
+    for (Point doc : main_dmset) {
+	dmset.add_document(doc.get_document());
+    }
+
+    vector<Point> diff_dmset = compute_diff_dmset(main_dmset);
```

Comment:
`diff_dmset`  isn't used here - I suspect `main_dmset` in the line below should be `diff_dmset` (or else it seems we are just adding the same documents twice).

Reply:
You're correct, I must've overlooked this. Fixed in a new commit.